lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -842,11 +1037,70 @@ def alerts(self) -> dict:
         return alerts
 
 
-class LokiPushApiConsumer(RelationManagerBase):
+class ConsumerBase(RelationManagerBase):
+    """Consumer's base class."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        relation_name: str = DEFAULT_RELATION_NAME,
+        alert_rules_path: str = DEFAULT_ALERT_RULES_RELATIVE_PATH,
+        allow_free_standing_rules: bool = False,
+    ):
+        super().__init__(charm, relation_name)
+        self._charm = charm
+        self._relation_name = relation_name
+        self.allow_free_standing_rules = allow_free_standing_rules
+        self._alert_rules_path = _resolve_dir_against_charm_path(charm, alert_rules_path)
+        self.topology = JujuTopology.from_charm(charm)
+
+    def _handle_alert_rules(self, relation):
+        if self._charm.unit.is_leader():
+            alert_groups, invalid_files = load_alert_rules_from_dir(
+                self._alert_rules_path,
+                self.topology,
+                recursive=False,
+                allow_free_standing=self.allow_free_standing_rules,
+            )
+            alert_rules_error_message = self._check_alert_rules(alert_groups, invalid_files)
+
+            if alert_rules_error_message:
+                self.on.loki_push_api_alert_rules_error.emit(alert_rules_error_message)
+
+            relation.data[self._charm.app]["metadata"] = json.dumps(self.topology.as_dict())
+            relation.data[self._charm.app]["alert_rules"] = json.dumps({"groups": alert_groups})
+
+    def _check_alert_rules(self, alert_groups, invalid_files) -> str:
+        """Check alert rules.
+
+        Args:
+            alert_groups: a list of prometheus alert rule groups.
+            invalid_files: a list of invalid rules files.
+
+        Returns:
+            A string with the validation message. The message is not empty whether there are
+            invalid alert rules files or there are no alert rules groups.
+        """
+        message = ""
+
+        if invalid_files:
+            must_contain = ["'alert'", "'expr'"]
+            if not self.allow_free_standing_rules:
+                must_contain.append("'%%juju_topology%%'")
+
+            message = "Failed to read alert rules (must contain {}): ".format(
+                ", ".join(must_contain)
+            ) + ", ".join(map(str, invalid_files))
+        elif not alert_groups:
+            message = "No alert rules found in {}".format(self._alert_rules_path)
```

Comment:
`%%juju_topology%%` is not a requirement. However, the consumer needs to strap on labels containing the juju topology, and the provider needs to read them and use `promql_transform` to inject them into the expression.

Reply:
I refactored all that in https://github.com/canonical/loki-k8s-operator/pull/79